src/data/sum/order.lean

```diff
@@ -0,0 +1,491 @@
+{ map_rel_iff' := by { rintro ((a | a) | a) ((b | b) | b); simp },
+  ..equiv.sum_assoc α β γ }
+
+@[simp] lemma sum_assoc_apply_in1 (a : α) : sum_assoc α β γ (inl (inl a)) = inl a := rfl
+@[simp] lemma sum_assoc_apply_in2 (b : β) : sum_assoc α β γ (inl (inr b)) = inr (inl b) := rfl
+@[simp] lemma sum_assoc_apply_in3 (c : γ) : sum_assoc α β γ (inr c) = inr (inr c) := rfl
+
+/-- `order_dual` is distributive over `⊕` up to an order isomorphism. -/
+def sum_dual_distrib (α β : Type*) [has_le α] [has_le β] :
+  order_dual (α ⊕ β) ≃o order_dual α ⊕ order_dual β :=
+{ map_rel_iff' := begin
+  rintro (a | a) (b | b),
+  { change inl (to_dual a) ≤ inl (to_dual b) ↔ to_dual (inl a) ≤ to_dual (inl b),
+    simp only [to_dual_le_to_dual, inl_le_inl_iff] },
+  { exact iff_of_false not_inl_le_inr not_inr_le_inl },
+  { exact iff_of_false not_inr_le_inl not_inl_le_inr },
+  { change inr (to_dual a) ≤ inr (to_dual b) ↔ to_dual (inr a) ≤ to_dual (inr b),
+    simp only [to_dual_le_to_dual, inr_le_inr_iff] }
+end,
+  ..equiv.refl _ }
+
+@[simp] lemma sum_dual_distrib_inl (a : α) :
+  sum_dual_distrib α β (to_dual (inl a)) = inl (to_dual a) := rfl
+
+@[simp] lemma sum_dual_distrib_inr (b : β) :
+  sum_dual_distrib α β (to_dual (inr b)) = inr (to_dual b) := rfl
+
+/-- `equiv.sum_assoc` promoted to an order isomorphism. -/
+def sum_lex_assoc (α β γ : Type*) [has_le α] [has_le β] [has_le γ] :
+  (α ⊕ₗ β) ⊕ₗ γ ≃o α ⊕ₗ β ⊕ₗ γ :=
+{ map_rel_iff' := begin
+    rintro ((a | a) | a) ((b | b) | b),
+    { change to_lex (inl a) ≤ to_lex (inl b) ↔
+        to_lex (inl $ to_lex $ inl a) ≤ to_lex (inl $ to_lex $ inl b),
+      simp only [lex.inl_le_inl_iff] },
+    { exact iff_of_true (lex.inl_le_inr _ _) (lex.inl_le_inl_iff.2 $ lex.inl_le_inr _ _) },
+    { exact iff_of_true (lex.inl_le_inr _ _) (lex.inl_le_inr _ _) },
+    { exact iff_of_false lex.not_inr_le_inl (λ h, lex.not_inr_le_inl $ lex.inl_le_inl_iff.1 h) },
+    { change to_lex (inr $ to_lex $ inl a) ≤ to_lex (inr $ to_lex $ inl b) ↔
+        to_lex (inl $ to_lex $ inr a) ≤ to_lex (inl $ to_lex $ inr b),
+      simp only [lex.inl_le_inl_iff, lex.inr_le_inr_iff] },
+    { exact iff_of_true (lex.inr_le_inr_iff.2 $ lex.inl_le_inr _ _) (lex.inl_le_inr _ _) },
+    { exact iff_of_false lex.not_inr_le_inl lex.not_inr_le_inl },
+    { exact iff_of_false (λ h, lex.not_inr_le_inl $ lex.inr_le_inr_iff.1 h) lex.not_inr_le_inl },
+    { change to_lex (inr $ to_lex $ inr a) ≤ to_lex (inr $ to_lex $ inr b) ↔
+        to_lex (inr a) ≤ to_lex (inr b),
+      simp only [lex.inr_le_inr_iff] }
+  end,
+  ..equiv.sum_assoc α β γ }
+
+@[simp] lemma sum_lex_assoc_apply_in1 (a : α) :
+  sum_lex_assoc α β γ (to_lex $ inl $ to_lex $ inl a) = to_lex (inl a) := rfl
+
+@[simp] lemma sum_lex_assoc_apply_in2 (b : β) :
+  sum_lex_assoc α β γ (to_lex $ inl $ to_lex $ inr b) = to_lex (inr $ to_lex $ inl b) := rfl
+
+@[simp] lemma sum_lex_assoc_apply_in3 (c : γ) :
+  sum_lex_assoc α β γ (to_lex $ inr c) = to_lex (inr $ to_lex $ inr c) := rfl
```

Comment:
Instead of `in1`, `in2`, `in3`, I think `inl_inl`, `inl_inr`, and `inr` would be better.

Reply:
Should I then change the corresponding `equiv.sum_assoc` lemmas?